File: lib/ts/src/internal/types/TAny.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { TTypeDeclaration } from './TTypeDeclaration.js';
import { ValidationFailure } from '../validation/ValidationFailure.js';
import { TType } from './TType.js';
import { generateRandomAny } from '../generation/GenerateRandomAny.js';
import { GenerateContext } from '../generation/GenerateContext.js';
import { ValidateContext } from '../validation/ValidateContext.js';

const anyName = 'Any';

export class TAny extends TType {
    getTypeParameterCount(): number {
        return 0;
    }

    validate(value: any, typeParameters: TTypeDeclaration[], ctx: ValidateContext): ValidationFailure[] {
        return [];
    }

    generateRandomValue(
        blueprintValue: any,
        useBlueprintValue: boolean,
        typeParameters: TTypeDeclaration[],
        ctx: GenerateContext,
    ): any {
        return generateRandomAny(ctx);
    }

    getName(): string {
        return anyName;
    }
}

File: lib/go/internal/types/TAny.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package types

const anyName = "Any"

// TAny represents Telepact's wildcard type.
type TAny struct{}

// NewTAny constructs a TAny instance.
func NewTAny() *TAny {
	return &TAny{}
}

// GetTypeParameterCount implements TType.
func (t *TAny) GetTypeParameterCount() int {
	return 0
}

// Validate implements TType by accepting all values.
func (t *TAny) Validate(value any, typeParameters []*TTypeDeclaration, ctx *ValidateContext) []*ValidationFailure {
	return nil
}

// GenerateRandomValue implements TType by delegating to GenerateRandomAny.
func (t *TAny) GenerateRandomValue(blueprintValue any, useBlueprintValue bool, typeParameters []*TTypeDeclaration, ctx *GenerateContext) any {
	return GenerateRandomAny(ctx)
}

// GetName implements TType.
func (t *TAny) GetName(ctx *ValidateContext) string {
	return anyName
}
